src/pybind/mgr/mgr_util.py

```diff
@@ -176,3 +178,109 @@ def verify_tls_files(cert_fname, pkey_fname):
+    :type rates: list[tuple[float, float]]
+
+    :return: The last derivative or 0.0 if none exists
+    :rtype: float
+
+    >>> get_most_recent_rate(None)
+
+    >>> get_most_recent_rate([(1, -2.0)])
+    -2.0
+    >>> get_most_recent_rate([(1, 2.0), (2, 1.5), (3, 5.0)])
+    5.0
+    """
+    if rates is None or len(rates) == 0:
+        return None
+    return rates[-1][1]
+
+def get_time_series_rates(data):
+    """ Rates from time series data
+
+    :param data: Time series data where the time is provided in seconds
+    :type data: list[tuple[float, float]]
+
+    :return: The derivative between all time series data points
+    :rtype: list[tuple[float, float]]
+
+    >>> logger.debug = lambda x: print(x)
+    >>> get_time_series_rates([])
+
+    >>> get_time_series_rates([[0, 1], [1, 3]])
+    [(1, 2.0)]
+    >>> get_time_series_rates([[0, 2], [0, 3], [0, 1], [1, 2], [1, 3]])
+    Invalid time series samples [[0, 2], [0, 3], [1, 2]] in [[0, 2], [0, 3], [0, 1], [1, 2], [1, 3]]
+    [(1, 2.0)]
+    >>> get_time_series_rates([[1, 1], [2, 3], [4, 11], [5, 16], [6, 22]])
+    [(2, 2.0), (4, 4.0), (5, 5.0), (6, 6.0)]
+    """
+    data = _filter_time_series(data)
+    if data is None:
+        return None
+    return [(data2[0], _derivative(data1, data2)) for data1, data2 in
+            pairwise(data)]
+
+def _filter_time_series(data):
+    """ Filters invalid samples from time series data
+    It also enforces the list to contain at least two samples.
+
+    All filtered values will be shown in the debug log. If values were filtered it's a bug in the
+    time series data collector, please report it.
+
+    :param data: Time series data where the time is provided in seconds
+    :type data: list[tuple[float, float]]
+
+    :return: Filtered time series data
+    :rtype: list[tuple[float, float]]
+
+    >>> logger.debug = lambda x: print(x)
+    >>> _filter_time_series([])
+
+    >>> _filter_time_series([[1, 42]])
+
+    >>> _filter_time_series([[10, 2], [10, 3]])
+    Invalid time series samples [[10, 2]] in [[10, 2], [10, 3]]
+
+    >>> _filter_time_series([[0, 1], [1, 2]])
+    [[0, 1], [1, 2]]
+    >>> _filter_time_series([[0, 2], [0, 3], [0, 1], [1, 2], [1, 3]])
+    Invalid time series samples [[0, 2], [0, 3], [1, 2]] in [[0, 2], [0, 3], [0, 1], [1, 2], [1, 3]]
+    [[0, 1], [1, 3]]
+    >>> _filter_time_series([[1, 1], [2, 3], [4, 11], [5, 16], [6, 22]])
+    [[1, 1], [2, 3], [4, 11], [5, 16], [6, 22]]
+    """
+    if data is None or len(data) < 2:
+        return None
+    last_index = len(data) - 1
+    isValid = lambda i, d: i == last_index or d[0] != data[i + 1][0]
+    filtered = [d for i, d in enumerate(data) if isValid(i, d)]
+    if len(filtered) <= last_index:
+        removed = [d for i, d in enumerate(data) if not isValid(i, d)]
+        logger.debug('Invalid time series samples {} in {}'.format(removed, data))
+    if len(filtered) < 2:
+        return None
+    return filtered
```

Comment:
I think the current implementation is still confusing. What do you think about changing it to the following implementation?

```python
filtered = []
for i in range(len(data)-1):
  if data[i][0] == data[i+1][0]:
    log.debug("Found duplicate sample with same timestamp: %s, %s", data[i], data[i+1])
    continue
  filtered.append(data[i])
filtered.append(data[-1])
return filtered
```

Reply:
Thanks, nice work :)